File: android/app/src/main/java/com/runformcoach/runformcoachai/WeeklyInsightModels.kt
package com.runformcoach.runformcoachai

import com.google.gson.annotations.SerializedName

// ── Backend Response Models ───────────────────────────────────────────────────

/**
 * Response from GET /sessions/trends — 4-week trend data.
 */
data class WeeklyTrendsResponse(
    @SerializedName("current_week") val currentWeek: WeekSummary,
    @SerializedName("previous_week") val previousWeek: WeekSummary,
    @SerializedName("weekly_trends") val weeklyTrends: List<WeekSummary> = emptyList(),
    @SerializedName("ai_suggestion") val aiSuggestion: String? = null,
    @SerializedName("badges") val badges: List<UserBadge> = emptyList()
)

/**
 * One week's aggregated running metrics.
 */
data class WeekSummary(
    @SerializedName("week_label") val weekLabel: String = "",
    @SerializedName("week_start_iso") val weekStartIso: String = "",
    @SerializedName("avg_cadence_spm") val avgCadenceSPM: Double = 0.0,
    @SerializedName("avg_amplitude_cm") val avgAmplitudeCm: Double = 0.0,
    @SerializedName("avg_gct_ms") val avgGCTMs: Double = 0.0,
    @SerializedName("total_distance_km") val totalDistanceKm: Double = 0.0,
    @SerializedName("total_sessions") val totalSessions: Int = 0,
    @SerializedName("total_duration_min") val totalDurationMin: Double = 0.0
)

/**
 * Achievement badge earned this week.
 */
data class UserBadge(
    @SerializedName("badge_id") val badgeId: String = "",
    @SerializedName("badge_name") val badgeName: String = "",
    @SerializedName("badge_icon") val badgeIcon: String = "",
    @SerializedName("badge_description") val badgeDescription: String = ""
)

// ── ViewModel State ───────────────────────────────────────────────────────────

/**
 * UI state for the weekly insight screen.
 */
sealed class WeeklyInsightState {
    object Loading : WeeklyInsightState()
    data class Success(val data: WeeklyTrendsResponse) : WeeklyInsightState()
    data class Error(val message: String) : WeeklyInsightState()
}

// ── Delta helpers ─────────────────────────────────────────────────────────────

/** Symbolic arrow direction for trend indicators. */
enum class TrendDirection { UP, DOWN, FLAT }

/**
 * Compute the delta and direction between this week and last week.
 */
data class MetricDelta(
    val currentValue: Double,
    val previousValue: Double,
    val delta: Double,
    val deltaPct: Double,
    val direction: TrendDirection,
    val unit: String
)

fun computeDelta(
    current: Double,
    previous: Double,
    unit: String,
    invertGood: Boolean = false  // true when lower is better (e.g. GCT, amplitude)
): MetricDelta {
    val delta = current - previous
    val deltaPct = if (previous != 0.0) (delta / previous) * 100.0 else 0.0
    val rawDirection = when {
        delta > 0.01 -> TrendDirection.UP
        delta < -0.01 -> TrendDirection.DOWN
        else -> TrendDirection.FLAT
    }
    val direction = if (invertGood) {
        when (rawDirection) {
            TrendDirection.UP -> TrendDirection.DOWN
            TrendDirection.DOWN -> TrendDirection.UP
            TrendDirection.FLAT -> TrendDirection.FLAT
        }
    } else rawDirection
    return MetricDelta(current, previous, delta, deltaPct, direction, unit)
}

File: ios/RunFormCoachAI/WeeklyInsightModels.swift
import Foundation

// MARK: - Backend Response Models

/// Response from GET /api/v1/sessions/trends — weekly trend data.
struct WeeklyTrendsResponse: Codable {
    let currentWeek: WeekSummary
    let previousWeek: WeekSummary
    let weeklyTrends: [WeekSummary]
    let aiSuggestion: String?
    let badges: [UserBadge]

    enum CodingKeys: String, CodingKey {
        case currentWeek = "current_week"
        case previousWeek = "previous_week"
        case weeklyTrends = "weekly_trends"
        case aiSuggestion = "ai_suggestion"
        case badges
    }
}

/// One week's aggregated running metrics.
struct WeekSummary: Codable {
    let weekLabel: String
    let weekStartIso: String
    let avgCadenceSPM: Double
    let avgAmplitudeCm: Double
    let avgGCTMs: Double
    let totalDistanceKm: Double
    let totalSessions: Int
    let totalDurationMin: Double

    enum CodingKeys: String, CodingKey {
        case weekLabel = "week_label"
        case weekStartIso = "week_start_iso"
        case avgCadenceSPM = "avg_cadence_spm"
        case avgAmplitudeCm = "avg_amplitude_cm"
        case avgGCTMs = "avg_gct_ms"
        case totalDistanceKm = "total_distance_km"
        case totalSessions = "total_sessions"
        case totalDurationMin = "total_duration_min"
    }
}

/// Achievement badge earned this week.
struct UserBadge: Codable, Identifiable {
    var id: String { badgeId }
    let badgeId: String
    let badgeName: String
    let badgeIcon: String
    let badgeDescription: String

    enum CodingKeys: String, CodingKey {
        case badgeId = "badge_id"
        case badgeName = "badge_name"
        case badgeIcon = "badge_icon"
        case badgeDescription = "badge_description"
    }
}

// MARK: - ViewModel State

enum WeeklyInsightState {
    case loading
    case success(WeeklyTrendsResponse)
    case error(String)
}

// MARK: - Delta Helpers

enum TrendDirection {
    case up, down, flat
}

struct MetricDelta {
    let currentValue: Double
    let previousValue: Double
    let delta: Double
    let deltaPct: Double
    let direction: TrendDirection
    let unit: String
}

func computeDelta(
    current: Double,
    previous: Double,
    unit: String,
    invertGood: Bool = false
) -> MetricDelta {
    let delta = current - previous
    let deltaPct = previous != 0 ? (delta / previous) * 100.0 : 0.0
    let rawDirection: TrendDirection = {
        if delta > 0.01 { return .up }
        if delta < -0.01 { return .down }
        return .flat
    }()
    let direction: TrendDirection = invertGood
        ? (rawDirection == .up ? .down : rawDirection == .down ? .up : .flat)
        : rawDirection
    return MetricDelta(
        currentValue: current,
        previousValue: previous,
        delta: delta,
        deltaPct: deltaPct,
        direction: direction,
        unit: unit
    )
}
